packages/smart-contracts/src/contracts/ProxyChangeCryptoFiat.sol

```diff
@@ -0,0 +1,127 @@
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+
+// Mock up for chainlink on private network
+import "./lib/MockPrivateChainlinkAggregatorCaller.sol";
+
+// Chainlink aggregator for rinkeby
+// import "./lib/RinkebyChainlinkAggregatorCaller.sol";
+
+// Chainlink aggregator for mainnet
+// import "./lib/MainnetChainlinkAggregatorCaller.sol";
+
+
+/**
+ * @title ProxyChangeCryptoFiat
+ */
+contract ProxyChangeCryptoFiat is ChainlinkAggregatorCaller {
+  using SafeMath for uint256;
+
+  // Event to declare a transfer with a reference
+  event TransferWithReference(
+    address tokenAddress,
+    address to,
+    uint256 amount,
+    bytes indexed paymentReference,
+    uint256 amountinCrypto
+  );
+
+  /**
+   * @notice Performs a ERC20 token transfer with a reference computing the amount regarding a fiat amount
+   * @param _to Transfer recipient
+   * @param _amountFiat Amount reference to transfer in fiat (8 decimals)
+   * @param _currencyFiat currency of the amount reference
+   * @param _currencyCrypto currency of the crypto to pay with
+   * @param _paymentReference Reference of the payment related
+   * @param _maxCryptoToSpend amount max that we can spend on the behalf of the user
+   */
+  function transferWithReference(
+    address _to,
+    uint256 _amountFiat,
+    FiatEnum _currencyFiat,
+    CryptoEnum _currencyCrypto,
+    bytes calldata _paymentReference,
+    uint256 _maxCryptoToSpend
+  ) external
+  {
+    // TODO !
+    require(_currencyCrypto != CryptoEnum.ETH, "ETH not supported yet");
+
+    // Get the amount to pay in the crypto currency chosen
+    uint256 amountToPay = computeConversion(_amountFiat, _currencyFiat, _currencyCrypto);
+    require(amountToPay <= _maxCryptoToSpend, "Amount to pay is over the user limit");
+
+    // Make the ERC20 transfer
+    IERC20 erc20 = IERC20(getTokenAddress(_currencyCrypto));
+    require(
+      erc20.allowance(msg.sender, address(this)) >= amountToPay,
+      "allowance needed for the token transfer"
+    );
+    require(erc20.transferFrom(msg.sender, _to, amountToPay), "transferFrom() failed");
```

Comment:
Isn't it possible to call the existing proxy contract to do do the transfer, let it throw the usual transfer with reference event, so the main blockchain payment detection is kept the same. This contract would send an event to match fiat amount and crypto amount. 

I think there are pros and cons for both kinds of solutions, but we need to consider them and decide.

Reply:
We should have an IRL discussion about it

Reply:
IRL
Tradeoff between:
- more info on blockchain
- or more calls to the blockchain nodes (Infura)

To be challenged